### ISSUE TITLE
Bug: LocalFree

### ISSUE THREAD

**Issue by jxy-s** [microsoft/win32metadata#1557]
### Which crate is this about?

windows

### Crate version

master

### Summary

`LocalFree` has incorrect error handling. `LocalFree` returns `NULL` on success. This will return an error when `LocalFree` succeeds:

https://github.com/microsoft/windows-rs/blob/2e7f99d9ca80bb8f1216db7964062daf42e80158/crates/libs/windows/src/Windows/Win32/System/Memory/mod.rs#L465-L472

**Comment by mikebattista** [comment]
Should this be marked with `[CanReturnAlternateSuccessCodes]` or `[CanReturnErrorsAsSuccess]`? We don't use preservesig.

**Comment by kennykerr** [comment]
Sure, I don't recall the difference between these two attributes.

**Comment by mikebattista** [comment]
https://github.com/microsoft/win32metadata/issues/1315

**Comment by kennykerr** [comment]
Either works for me. I just treat them all as "don't transform signature" in Rust, but I guess "CanReturnErrorsAsSuccess" sounds like its closest to the intent here. @riverar may have a better answer as he invented this distinction.

**Comment by riverar** [comment]
This API returns a handle, not an error code, and we annotate the invalid handle values. So it's not clear what metadata changes are needed here. Looks like a crate bug?

```
public static extern HLOCAL LocalFree([Optional][In] HLOCAL hMem);

[RAIIFree("LocalFree")]
[InvalidHandleValue(-1L)]
[InvalidHandleValue(0L)]
[NativeTypedef]
public struct HLOCAL
{
	public IntPtr Value;
}

```

**Comment by kennykerr** [comment]
See https://github.com/microsoft/windows-rs/issues/2392#issuecomment-1477765781

**Comment by riverar** [comment]
I see, this API is a mutant and returns invalid handles on success. That's unfortunate.

Agree, `[CanReturnErrorsAsSuccess]` is the appropriate attribute here. Also agree developers should ignore the return value as they can detect for invalid handles earlier and that's the only useful return out of this thing.

**Comment by jxy-s** [comment]
At times you don't know the handle is invalid or corrupt until you pass it to an API to do something with it. The user of the API has no insight, really, into the internal handle table.

**Comment by jxy-s** [comment]
Also, I hate to be pedantic here, but accuracy for this project matters. And technically... `0` is a "valid handle" to pass per the API contract. As the SAL implies `_Frees_ptr_opt_` it's optional, and a `NULL` value is safe and is documented as such:

>If the hMem parameter is NULL, LocalFree ignores the parameter and returns NULL.

**Comment by jxy-s** [comment]
Let's consider it in the abstract. 

Any handle on the system is subject to being manipulated by another process or the kernel. A handle in a process _can be closed_ by a separate process that isn't the one who owns it.

Or, an even simpler case of a closing a handle twice.

These are situations where error codes should be propagated clearly such that bugs and not receiving a correct error code can be detrimental for visibility and debug-ability of any system.

**Comment by riverar** [comment]
`HLOCAL` is just a `HANDLE` typedef therefore it logically inherits the same _invalid value_ characteristics. That is, `0` and `-1` are both integers used to represent invalid handles. (`0` is sourced from `NULL`, while `-1` is covered by `INVALID_HANDLE_VALUE`.) We encode that into the metadata and the Rust crate generates a helper `.is_invalid() -> bool` function that developers can use to check if their `HANDLE` or `HLOCAL` is invalid _as it sits_.

> Also, I hate to be pedantic here, but accuracy for this project matters. And technically... 0 is a "valid handle" to pass per the API contract. As the SAL implies _Frees_ptr_opt_ it's optional, and a NULL value is safe and is documented as such

Not quite. Acceptance of invalid handle values does not make the handles valid. It just means the API accepts invalid handles and handles them in the documented manner.

**Comment by jxy-s** [comment]
Thanks for taking care of this issue @mikebattista.

@riverar I think my point is being overlooked, I may have done a poor job at expressing it.
> Not quite. Acceptance of invalid handle values does not make the handles valid. It just means the API accepts invalid handles and handles them in the documented manner.

Yes, that's what I said. Perhaps the quotes didn't read the way I intended them to... 

The point I'd like to make clear: SAL an excellent source of documentation (usually better than MSDN). Everything you need to know about the input and output of the API is right on the signature. And from what I've gathered the metadata doesn't take that into account. _It should._

Regardless, I appreciate everyone's effort. Thank you.

**Comment by mikebattista** [comment]
We process SAL annotations that ClangSharp scrapes. It doesn't scrape them in this case.

@tannergooding FYI.

**Comment by riverar** [comment]
> SAL an excellent source of documentation (usually better than MSDN) [...] And from what I've gathered the metadata doesn't take that into account. It should.

Agreed. We do some minimal handling of some SAL variants in win32metadata, but there's certainly room for improvement. I believe clangsharp spits out all the raw data we need, we _just_ need to build a SAL lexer/parser.

**Comment by mikebattista** [comment]
No it's not scraping `_Success_(return==0)`.

**Comment by 0vercl0k** [comment]
I stumbled upon this issue as I ran into the same behavior that jxy-s reported (thank you!) but according to the above conversations; it seems like this is fixed?
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: HRESULT(0x00000000), message: "The operation completed successfully." }', src\thread.rs:56:62
```

I ran into this with the latest `windows` crate (`v0.51.1`) - am I doing something wrong / missing something? 🤔

Cheers

**Comment by 0vercl0k** [comment]
That makes sense - thanks for the details, I'll wait patiently for a new
release :)

Cheers

**Comment by riverar** [comment]
@0vercl0k I think in the interim, we can get that merged into the master branch soon (today?) so you can point to the repository in your Cargo.toml.

**Comment by 0vercl0k** [comment]
I'd love that - if you end up making this happen let me know and I'll be
happy to try it out!

Cheers

**Comment by riverar** [comment]
Ah, Rust is currently blocked waiting for https://github.com/microsoft/win32metadata/commit/1b2ffca90f23760c8ff339756d6483b0f919f9ab to make its way to a release.

cc: @mikebattista for awareness

**Comment by 0vercl0k** [comment]
I see, bummer :( Thank you for letting me know!

Cheers

**Comment by riverar** [comment]
@mikebattista The latest release has the LsaRegisterLogonProcess change (https://github.com/microsoft/win32metadata/issues/1735) that was reverted hours later that didn't make it into that release. So we don't want to release a new Rust crate with that bug in it.

**Comment by mikebattista** [comment]
There's no further changes there. It was reverted by design since Reserved wasn't appropriate.

**Comment by riverar** [comment]
@mikebattista Ah, there was some confusion here because the release notes still have the change listed. So it didn't look like the revert made it in.

But looking at GitHub, it did! https://github.com/microsoft/win32metadata/commits/v56.0.13-preview

We'll get the wheels spinning on the Rust side, thanks for the ping.

**Comment by mikebattista** [comment]
Excellent. Glad you're unblocked.

**Comment by 0vercl0k** [comment]
Rafael, let me know if there is / when there is an issue or something I can
subscribe to to follow the work :)

Thank you all!

**Comment by 0vercl0k** [comment]
Amazing, thank you so much Rafael 🙏

**Comment by riverar** [comment]
Reopening. `[CanReturnErrorsAsSuccess]` is not used by Rust. This attribute is used to markup APIs that return errors during common usage. (This would result in exceptions on in C#, for example.) LocalFree has an abnormal form here and should really be marked `[PreserveSig]` so that all language projections treat it the same.

**Comment by mikebattista** [comment]
We had a long discussion about these attributes and PreserveSig at https://github.com/microsoft/win32metadata/issues/1315.

I'm not sure PreserveSig is correct here based on that discussion.

**Comment by riverar** [comment]
@mikebattista In this case, the API doesn't return an error code or HRESULT, but rather a handle, so I think both of the attributes are off the table. This one appears to need real `PreserveSig` treatment.

(Recall `[CanReturnMultipleSuccessValues]` is for APIs that return S_FALSE, S_OK, etc. which requires special handling in Rust, while `[CanReturnErrorsAsSuccess]` is for APIs that return errors that could throw in C#, like DXGI_ERROR_NOT_FOUND, during the normal course of operations.)

**Comment by tannergooding** [comment]
This should already be handled by clangsharp via the `NativeAnnotation` emissions, so if you define `-D_PREFAST_=1` then you should get `[NativeAnnotation("Name=SAL_success; p1=return == 0")]`

**Comment by kennykerr** [comment]
This is no longer an issue for Rust now that we're generating our own metadata directly from the Windows SDK headers.
